spi_nxp_lpspi: RX bigger than TX read all RX

decsny opened the pull request:
Change the driver behavior so that if the provided RX buffer set is bigger than the TX buffer, we will read all the RX buffer and fill the TX with NOPs. The SPI driver API does not say to do this, and my original interpretation of the API was that the TX length controls the entire transfer length, but this behavior might fit better with some de facto expectations of in tree consumers.

Alternative to #85526 to Fixes #85355

There is clearly some ambiguity in the SPI API, I have on my todo list to improve the documentation of SPI API, and probably this behavior in this PR is preferable.

Fixes #85901

I also plan to add some test cases to the SPI loopback test suite soon to catch many of these bugs LPSPI is having that are not caught by the test.

Commit:
spi_nxp_lpspi: RX bigger than TX read all RX

Change the driver behavior so that if the provided RX buffer set is
bigger than the TX buffer, we will read all the RX buffer and fill the
TX with NOPs. The SPI driver API does not say to do this, and my
original interpretation of the API was that the TX length controls the
entire transfer length, but this behavior might fit better with some de
facto expectations of in tree consumers.

Also add some robustness to the calculation of how many extra bytes to fill
when tx should be nops.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented:
> Tried this PR and it does fix the issue reported. However, I do see an extra byte clocked out in the transfer: ![image](https://private-user-images.githubusercontent.com/65253049/411697493-8d749a98-5f95-4e35-8824-c4f319904fd8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyMjUzNjksIm5iZiI6MTczOTIyNTA2OSwicGF0aCI6Ii82NTI1MzA0OS80MTE2OTc0OTMtOGQ3NDlhOTgtNWY5NS00ZTM1LTg4MjQtYzRmMzE5OTA0ZmQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEwVDIyMDQyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE5ZmQ2OTk1Y2ZiNzhmZDhkY2E2ZDAzN2FmZDIzYTJmMWRlNTdhYWNjNzdiZTg2ZDhiNzU3Mjk2MzcwMDJmYmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pg--Oj350FSroiczj4kJcTe3Ztr4-MSp33HM6dIoNh8)

updated

jeppenodgaard commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi.c:
```diff
@@ -235,10 +201,16 @@ static void lpspi_isr(const struct device *dev)
 	}
 
-	if (!spi_context_tx_on(ctx) && lpspi_is_rx_done(dev)) {
+	if (!spi_context_tx_on(ctx) && spi_context_rx_len_left(ctx) == 1) {
```
`!spi_context_tx_on(ctx)` is checked up to 3 times. Move it out of the `if`s and add e.g.:
```
if (spi_context_tx_on(ctx)) {
	return;
}
```

azmat-bilal commented:
i just checked the latest changes on imxrt1170evkb and adxl372 sensor. the spi read operation is working but in write operation, the CS is deasserted after the first byte instead of the second. here is the timing diagram. the rea marked in blue is reading a register and the area in green is writing into it.

![image](https://github.com/user-attachments/assets/f7010234-69a3-4fb3-b6ef-bc3d7b166901)

decsny commented:
If you are using the spi_write function (as the adx driver does), then the expected behavior from this lpspi for the adx reg write with the reworked ISR in this PR driver is:
- transceive functions submits first buf (command byte probably?) to tx fifo
- interrupt happens, updates the context from 1st byte being sent and submits 2nd buf (len 1 byte data ?) to tx fifo
- tx still on so return
- rx is not at play for entire transfer since it is off always with NULL buf
- 2nd interrupt happens and updates context, which turns off since all tx was sent, isr checks both tx and rx are off, and since they are, it ends the transfer by writing TCR=0 which deassert CS. At this point all the data should have been sent, so don't know how is it that the CS deasserted before the SCK finished. 


What might help to know is if you are using the GPIO CS?

Maybe the IP interrupts before the spi finishes transceiving the word onto the bus. If that is the case, then I didn't know this. If it does, then GPIO CS could deassert before spi finishes, in which case maybe we need to postpone the gpio cs deassert. And maybe native too could have this problem?, but I doubt it because the command goes in the same fifo as the TX data exactly for this reason, to have a flexible transfer control.

azmat-bilal commented:
> If you are using the spi_write function (as the adx driver does), then the expected behavior from this lpspi for the adx reg write with the reworked ISR in this PR driver is:
> 
> * transceive functions submits first buf (command byte probably?) to tx fifo
> * interrupt happens, updates the context from 1st byte being sent and submits 2nd buf (len 1 byte data ?) to tx fifo
> * tx still on so return
> * rx is not at play for entire transfer since it is off always with NULL buf
> * 2nd interrupt happens and updates context, which turns off since all tx was sent, isr checks both tx and rx are off, and since they are, it ends the transfer by writing TCR=0 which deassert CS. At this point all the data should have been sent, so don't know how is it that the CS deasserted before the SCK finished.
> 
> What might help to know is if you are using the GPIO CS?
> 
> Maybe the IP interrupts before the spi finishes transceiving the word onto the bus. If that is the case, then I didn't know this. If it does, then GPIO CS could deassert before spi finishes, in which case maybe we need to postpone the gpio cs deassert. And maybe native too could have this problem?, but I doubt it because the command goes in the same fifo as the TX data exactly for this reason, to have a flexible transfer control.

yes im using gpio cs.

decsny commented:
> i did some more debugging and this may help you. i think the issue is how the TDF (transmit data flag) is used. what i think is happening is that the tdf flag ( and as a result the lpspi_isr) gets triggered not at the 8th clock bit but 2nd, after the tx has started. in the image below, i force toggled the cs line in the isr to see where the isr gets called and as you can see the isr gets called after 2 clocks. as a result, after sending the second byte, the tdf gets set and the isr gets triggered before the full 8bit clock ticks, and since no data needs to be sent it deasserts the cs.
> 
> ![image](https://private-user-images.githubusercontent.com/28139445/412042882-71c0f838-bd7b-42b2-a5e7-31d83d094d84.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyOTMzNTYsIm5iZiI6MTczOTI5MzA1NiwicGF0aCI6Ii8yODEzOTQ0NS80MTIwNDI4ODItNzFjMGY4MzgtYmQ3Yi00MmIyLWE1ZTctMzFkODNkMDk0ZDg0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjExVDE2NTczNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc5ZWYxMGJiMmUwYTAxOTEwZGQxMTcwMzBiY2I3NDliNmM0ZjRmZjM2N2RjY2NkMDNmYmVlY2EwMzEzNzU2MDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.jxc_tmG7ZPWwkomztbk6s_I72vXa8kndYoqThVhhFaw)
> 
> this actually makes sense since in the reference manual the TDF flag indicates that the data can be written to the transmit fifo.
> 
> i think instead of using tdf for end of tranmission, other flags like FCF or TCF can be used.
> 
> adding flag description
> 
> ![image](https://private-user-images.githubusercontent.com/28139445/412045299-12fbe94f-2513-4cfa-949b-b49689928069.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyOTMzNTYsIm5iZiI6MTczOTI5MzA1NiwicGF0aCI6Ii8yODEzOTQ0NS80MTIwNDUyOTktMTJmYmU5NGYtMjUxMy00Y2ZhLTk0OWItYjQ5Njg5OTI4MDY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjExVDE2NTczNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVkYTFhMDYwYTYxZTgwMmVmODAwOGMzNjZhNDhhMjA5NjE5N2NkNzQ1ZWRhYmRkZGUxNzhkOTMzNjJiZWUyM2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.jTLjI5OWpcyl-DKOW6GEW8oMbkWgEjLxGGUQFD4IvWo)

The other flags _cannot_ be used , for the reason I said above, we are not using the hardware frame size to delimit the transfer, we are using the continuous transfer. The TDF just is when the TX watermark is reached (which in this driver is set to 0, so that we send the whole designated fill length before getting the interrupt). Changing to use a different flag would require rearchitecting  the entire implementation.

> yes im using gpio cs.

If your pin for the gpio CS is the same as the native PCS pin, try switching to native. I have a feeling it will work better. I would like to know if this issue in this case is only affecting the GPIO CS because that might tell us something about whether the interrupt is happening before or after the last word is clocked out. If before, that would explain the behavior, this is my current theory that I have without access to a signal analyzer to debug at the moment

azmat-bilal commented:
i did some more testing and found out that the cs getting de-asserted during the write operation can happen after full second byte transfer in some cases. so there is no fixed pattern when the cs will get deasserted but the end results is that the write operations most of the time fail due to this. i will test the native pcs and see if that solves the issue.

ubieda reviewed (dismissed):
I confirm this fixes the regression (using CS GPIO was working before: is working with this PR) + the extra byte is not clocked out anymore.
![image](https://github.com/user-attachments/assets/7782eab9-ade6-467e-9985-04b0cfff41bf)
Also, this seems to work as well for PCS.
![image](https://github.com/user-attachments/assets/b4520e67-d9a9-459d-9c35-463756c5be92)

ubieda commented:
> i did some more testing and found out that the cs getting de-asserted during the write operation can happen after full second byte transfer in some cases. so there is no fixed pattern when the cs will get deasserted but the end results is that the write operations most of the time fail due to this. i will test the native pcs and see if that solves the issue.

@azmat-bilal could you provide some repro steps on the issue you're running into? I should be able to replicate them on my end.

azmat-bilal commented:
> > i did some more testing and found out that the cs getting de-asserted during the write operation can happen after full second byte transfer in some cases. so there is no fixed pattern when the cs will get deasserted but the end results is that the write operations most of the time fail due to this. i will test the native pcs and see if that solves the issue.
> 
> @azmat-bilal could you provide some repro steps on the issue you're running into? I should be able to replicate them on my end.

yeah sure. so i am basically using the adxl372 sensor interfaced to lpspi1 using mimxrt1170-evkb. im using the accel polling example. the overlay im using is

```
&pinctrl {

	pinmux_lpspi1: pinmux_lpspi1 {
		group0 {
			pinmux = 
					<&iomuxc_gpio_ad_07_gpio9_io06>,
					<&iomuxc_gpio_ad_28_lpspi1_sck>,
					<&iomuxc_gpio_ad_31_lpspi1_sdi>,
					<&iomuxc_gpio_ad_30_lpspi1_sdo>,
					<&iomuxc_gpio_ad_06_gpio9_io05>,
					<&iomuxc_gpio_ad_14_gpio9_io13>;
			drive-strength = "high";
			slew-rate = "fast";
		};
	}; 
};

&lpspi1 {
	status = "okay";
	cs-gpios =  <&gpio9 6 GPIO_ACTIVE_LOW>;
	pinctrl-0 = <&pinmux_lpspi1>;
    pinctrl-names = "default";
	/delete-property/ dmas;
	/delete-property/ dma-names;

	adxl372_imu:adxl372@0 {
		reg = <0>;
		compatible = "adi,adxl372";
		status = "okay";
		int1-gpios = <&gpio9 5 GPIO_ACTIVE_LOW>;
		spi-max-frequency = <1000000>;
	};
};
```
with these settings, if i use any of the spi_reg_write_mask() function while basically is a read and then write of a register, the read operation works but for the write operation the cs goes high prematurely. the the issue can be seem in the image below.
![image](https://github.com/user-attachments/assets/898c8e6c-3c6b-4d70-a726-62c4eaf79282)

azmat-bilal commented:
@decsny with the native cs, the issue is not there which makes sense since the cs should get deasserted after all the 8bits are clocked out.

decsny commented:
> @decsny with the native cs, the issue is not there which makes sense since the cs should get deasserted after all the 8bits are clocked out.

I pushed an update to the PR that I think might fix the GPIO CS as well

azmat-bilal commented:
I just discovered another issue. in the read operation, the NOPs in the tx fifo are filled incorrectly, resulting in additional bytes being read. e.g. i wanted to read 9 bytes from sensor registers and was getting 25 byte operation as seen below. 
![image](https://github.com/user-attachments/assets/151850ff-638c-4a3a-b64e-1e08c33e63bf)
the yellow trace is the rxdata ready call in the isr. 

the issue is occuring because of the way nops are filled:
```
if (spi_context_rx_len_left(ctx) == 1) {
	base->TCR = 0;
} else if (spi_context_rx_on(ctx)) {
lpspi_data->fill_len = MIN(ctx->rx_len, config->rx_fifo_size);
lpspi_fill_tx_fifo_nop(dev);
}
```
lpspi_fill_tx_fifo_nop(dev) gets called on every byte read interrupt and puts in additional nops in the tx buffer. 

i tested a quick solution to validate the assumption by getting the total NOPs to write at the beginning by using:
`lpspi_data->tx_total_nop = lpspi_data->rx_total_len - lpspi_data->tx_total_len;`

and then calling the lpspi_fill_tx_fifo_nop() only once:

```
if (spi_context_rx_len_left(ctx) == 1) {
		base->TCR = 0;
} else if (spi_context_rx_on(ctx)) {
	if (lpspi_data->tx_total_nop > 0) {
		lpspi_data->fill_len = lpspi_data->tx_total_nop;
		lpspi_fill_tx_fifo_nop(dev);
		lpspi_data->tx_total_nop = 0;
	}
}
```

this is however a very rough implementation and i think there are better ways to do it. but the results shown below does show that the current implementation is writing extra bytes.

![image](https://github.com/user-attachments/assets/de95f23f-1c0e-4194-ae37-104437e71c3a)

decsny commented:
@azmat-bilal I put some code to avoid putting extra bytes in the tx fifo if there are already some

ubieda reviewed (dismissed):
Refreshing my +1. I just re-tested and SPI_RTIO is also working as well.

manuargue commented:
`west build -b mr_canhubk3 tests/drivers/spi/spi_loopback/` gives:
```
*** Booting Zephyr OS build v4.0.0-5466-g164f42f862a2 ***
Running TESTSUITE spi_loopback
===================================================================
START - test_spi_loopback
I: SPI test on buffers TX/RX 0x204040a0/0x20404080, frame size = 8
I: SPI test slow config
I: Start complete multiple
I: Passed
I: Start complete loop
I: Passed
I: Start null tx
I: Passed
I: Start half start
I: Passed
I: Start half end
I: Passed
I: Start every 4
I: Passed
I: Start rx bigger than tx
E: Buffer contents are different: 0x30,0x31,0x32,0x33,0x34,0x35,0x36,0x37,
E:                            vs: 0x2d,0x31,0x32,0x33,0x34,0x35,0x36,0x37,

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/spi/spi_loopback/src/spi.c:506: spi_rx_bigger_than_tx: (1 is true)
Buffer contents are different
```

decsny commented:
> `west build -b mr_canhubk3 tests/drivers/spi/spi_loopback/` gives:
> 
> ```
> *** Booting Zephyr OS build v4.0.0-5466-g164f42f862a2 ***
> Running TESTSUITE spi_loopback
> ===================================================================
> START - test_spi_loopback
> I: SPI test on buffers TX/RX 0x204040a0/0x20404080, frame size = 8
> I: SPI test slow config
> I: Start complete multiple
> I: Passed
> I: Start complete loop
> I: Passed
> I: Start null tx
> I: Passed
> I: Start half start
> I: Passed
> I: Start half end
> I: Passed
> I: Start every 4
> I: Passed
> I: Start rx bigger than tx
> E: Buffer contents are different: 0x30,0x31,0x32,0x33,0x34,0x35,0x36,0x37,
> E:                            vs: 0x2d,0x31,0x32,0x33,0x34,0x35,0x36,0x37,
> 
>     Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/spi/spi_loopback/src/spi.c:506: spi_rx_bigger_than_tx: (1 is true)
> Buffer contents are different
> ```

Is that new from this PR?

bperseghetti reviewed (dismissed):
I'm temporarily removing my approve till we get clarity on what commit @manuargue was on when testing this PR.

decsny commented:
If the LPSPI on s32 has yet another different behavior, I do not have that hardware to debug it

bperseghetti commented:
> If the LPSPI on s32 has yet another different behavior, I do not have that hardware to debug it

@igalloway can we confirm if we have spare S32 hardware at Austin offices?

igalloway commented:
> > If the LPSPI on s32 has yet another different behavior, I do not have that hardware to debug it
> 
> @igalloway can we confirm if we have spare S32 hardware at Austin offices?

I'm assuming S32K3xx. Yes there's lots of inventory and several MR-B3RB-M robot buggies in the lab.

manuargue commented:
> Is that new from this PR?

It's reproducible with current HEAD of this branch (you can see the sha from Zephyr's boot banner in my previous msg), and also with:
ff4dc62f0e52 HEAD main
62b911feea9 spi_nxp_lpspi: Rewrite driver, fix native chip sel

But it works with the older version of the driver, so it's not new from this PR:
0e4a5ac6cbc drivers: audio: dmic: remove compatible from nxp,dmic child binding
```
*** Booting Zephyr OS build v4.0.0-4333-g0e4a5ac6cbcb ***
Running TESTSUITE spi_loopback
===================================================================
START - test_spi_loopback
I: SPI test on buffers TX/RX 0x204040a0/0x20404080, frame size = 8
I: SPI test slow config
I: Start complete multiple
I: Passed
I: Start complete loop
I: Passed
I: Start null tx
I: Passed
I: Start half start
I: Passed
I: Start half end
I: Passed
I: Start every 4
I: Passed
I: Start rx bigger than tx
I: Passed
I: Start complete large transfers
I: Passed
I: SPI test fast config
I: Start complete multiple
I: Passed
I: Start complete loop
I: Passed
I: Start null tx
I: Passed
I: Start half start
I: Passed
I: Start half end
I: Passed
I: Start every 4
I: Passed
I: Start rx bigger than tx
I: Passed
I: Start complete large transfers
I: Passed
I: Start complete loop
I: Passed
I: Start complete loop
I: Passed
I: All tx/rx passed
 PASS - test_spi_loopback in 0.209 seconds
===================================================================
TESTSUITE spi_loopback succeeded
```

decsny commented:
@azmat-bilal can you help to confirm if you're still having issues or not with the gpio CS and the extra TX nops on the latest version of the PR? 

the S32 seems like it was broken before this PR, so the only thing holding back this PR is those two issues. I'd like to make sure your use case is working before merging this

azmat-calmfloor commented:
i just checked the latest changes implemented and i am seeing 4 and sometimes 5 bytes being read additionally when a 9 byte read operation is done on an accelerometer (testing with adxl355 at the moment). the trace below shows it. 

![image](https://github.com/user-attachments/assets/61604163-b0a9-4f64-99b6-edff2f3f68b7)

with the rough changes i mentioned earlier, the correct number of bytes are read as shown below.

![image](https://github.com/user-attachments/assets/7582b120-5b38-41f1-8b7d-8102c05526ae)

one more thing to mention, this issue is not covered by the spi loopback test as the read operation ends when the right amount of bytes are read even though there are additional NOPs being clocked out. and it may not be an issue in most of the cases where the desired number of bytes are read and additional reads are ignored. however when using sensor FiFo, the additional reads move the sensor fifo pointer resulting in missed sensor data.

Commit:
spi_nxp_lpspi: Fix S32 regressions

There are two bugs that caused regression for S32:

First there is a silicon errata specifically for the mask version on
this board that causes FIFO flush to not work as expected. The
workaround is to do a module reset before each transfer.

Second there was a division error for word size > 1 byte. The division
should be rounded up, not down, otherwise there will be an infinite
interrupt loop because the TX fifo will not be written to but the TDR
interrupt enabled causes interrupt when TX fifo is empty.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
spi_nxp_lpspi: Fix Kconfig description

The Kconfig description is wrong, this driver can be used with or
without RTIO. Also, rename the kconfig to be less confusing, as _NORMAL
is meaningless, this is the CPU/interrupt based driver.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
spi_nxp_lpspi: Remove unused struct fields

These two fields were not being used and were a relic of some
intermediate implementation, and forgot to be removed.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented:
@azmat-bilal I still do not know how to reproduce your issue, but I made some debug logs for myself where I saw that sometimes, when getting to the part of the code that calculates how many tx nops to fill, there was actually already some new RX data in the RX fifo, meaning that perhaps the max fill should not just be the ctx->rx_len but less than that. So I added some even more code to account for this in the calculation.

azmat-bilal commented:
@decsny I pulled the latest changes and did some testing and have some interesting results.
to keep things simple, i modified the spiloop test  and added this section which basically mimics what the sensor drivers do for register read operation (write 1byte addr and read n bytes). 

```
static int spi_tx_addr_and_rx_data(struct spi_dt_spec *spec)
{
	const uint32_t tx_buf_size = 1;

	const struct spi_buf buf[2] = {
		{
			.buf = buffer_tx,
			.len = tx_buf_size
		},
		{
			.buf = buffer_rx,
			.len = BUF_SIZE
		}
	};

	struct spi_buf_set tx = {
		.buffers = buf,
		.count = 1
	};

	struct spi_buf_set rx = {
		.buffers = buf,
		.count = 2
	};

	LOG_INF("Start tx addr and rx data with guard bytes");

	(void)memset(buffer_rx, 0xff, BUF_SIZE);

	int ret = spi_transceive_dt(spec, &tx, &rx);
	if (ret) {
		LOG_ERR("Code %d", ret);
		zassert_false(ret, "SPI transceive failed");
		return -1;
	}

	return 0;
}
```
using this i scoped the spi lines and with the latest code the data looked good as shown below.

![image](https://github.com/user-attachments/assets/5353f9dc-ffd6-4029-975c-0b0cd8fabffc)

however the sensor i had used SPI mode 0 (CPO=0, CPH=0) where as the spiloop back is using spi mode 3. when i just change the spi mode to mode 0, i see the extra byte clocked out as shown below.

MODE 0
```
#define SPI_OP(frame_size) SPI_OP_MODE_MASTER  | MODE_LOOP  \
				| SPI_WORD_SET(frame_size) | SPI_LINES_SINGLE
```

MODE 3
```
#define SPI_OP(frame_size) SPI_OP_MODE_MASTER  | MODE_LOOP | SPI_MODE_CPOL \
	          | SPI_MODE_CPHA | SPI_WORD_SET(frame_size) | SPI_LINES_SINGLE
```

![image](https://github.com/user-attachments/assets/5edc154d-0680-4718-afef-ab84db8ca878)

with the code shared above you may try this as well. i did look at the spi data for the rest of the spiloop back tests and those look good though. so i believe the convoluted spi buffer logic for sensor readout could be causing this extra byte.

 the more than 1 byte nop issue seems to be resolved. ill try to modify the sensor readout code to use separate buffers for tx and rx and see how it looks. but i think the underlying spi driver is now working properly.

i still have to check the gpio based cs but the native cs is working reliably.

dleach02 commented:
@tbursztyka I've added myself on the Assignees list to help move this along for RC2. The changes are totally encapsulated within NXP.

If you disagree then go ahead and drop me from the assignee list and give this PR a review.

tbursztyka commented:
> If you disagree then go ahead and drop me from the assignee list and give this PR a review.

Told you it's fine. If I take too long to review, go ahead. Moreover if it's about a driver, and not about the API itself.

(Btw, rhe only way I get notified (through the app) from GitHub is to bring my name into a comment (like you just did). Else I get all the pr and comments by mail. It's completely flooded.)

Well now that I am here I'll have a quick look

tbursztyka reviewed (approved):
Lgtm
